### DISCUSSION TITLE
How to pipe?

### DISCUSSION THREAD

**Discussion by yamcodes** [radashi-org/radashi#217] [Q&A] [answered]
Hey team, comparable libraries provide the option to pipe operations, i.e.:
```ts
const DATA = [
  { name: "john", age: 20, gender: "m" },
  { name: "marry", age: 22, gender: "f" },
  { name: "samara", age: 24, gender: "f" },
  { name: "paula", age: 24, gender: "f" },
  { name: "bill", age: 33, gender: "m" },
];
_(DATA)
  .filter((x) => x.gender === "f")
//               ^ 'x' should be typed properly as a Record
  .groupBy((x) => x.age)
//                ^ same here
  .value();
```

1. Can this be achieved with the **Radashi** utils?
2. If the answer to (1) is "Yes", is it possible to achieve this in a **typesafe** way? (See comments)

I am familiar with [chain](https://radashi.js.org/reference/curry/chain/) and [compose](https://radashi.js.org/reference/curry/compose/) however all the examples point to custom functions, as opposed to the first party **Radashi** functions. I suspect this is because **Radashi** always expects the data as an input.

**Comment by aleclarson** [comment, marked as answer]
Hey @yamcodes, thanks for bringing this up. Your example appears to be an anti-pattern, since tree-shaking wouldn't be possible. However, I'd be open to a PR that makes it easier to combine Radashi functions with `chain` and `compose` (presumedly by re-arranging the data arguments to be last in an optional `radashi/fp` submodule).

Alternatively, there are many popular, actively maintained, FP-focused libraries to choose from, like [fp-ts](https://gcanti.github.io/fp-ts/) or [ramda](https://ramdajs.com/) (though I can't vouch for their quality). As such, it's not an explicit goal of Radashi to compete in that arena.

**Reply by yamcodes** [reply]
Hey @aleclarson, thank you for your response and for setting the foundation for great collaborative work in this project.

You're absolutely right. The example is an "anti-pattern" in the realm of frontend, and everything we're trying to leave behind in the days of Lodash.
However, as you hinted, there are ways to make it tree shakeable. The example above can be reimagined as:

```ts
import { filter, groupBy, pipe } from 'my-awesome-utilities`;
const DATA = [
  { name: "john", age: 20, gender: "m" },
  { name: "marry", age: 22, gender: "f" },
  { name: "samara", age: 24, gender: "f" },
  { name: "paula", age: 24, gender: "f" },
  { name: "bill", age: 33, gender: "m" },
];

pipe(
  DATA,
  filter((x) => x.gender === "f"),
  groupBy((x) => x.age),
);
```

This can be achieved by making the functions "data-last", like you said.

As such we're aligned on your idea for said PR. I'd love to make time and contribute. Thank you!

**Reply by aleclarson** [reply]
> As such we're aligned on your idea for said PR. I'd love to make time and contribute.

That's great to hear! We love supporting new contributors.

Thinking more about it, the implementation would need to be done in such a way that doesn't add significant maintenance burden. To accomplish that, I think using “code generation” would be the right approach, so we can run a script to regenerate the parts of `radashi-fp` that would otherwise be a chore to keep up-to-date. Using code generation would probably mean using the `ts-morph` npm package. If you're not experienced with code generation, you could try consulting with ChatGPT or similar for guidance. Of course, I'll be here to help review your work and give advice. 👍
